### PR TITLE
Some fixes and improvements for "go to all" search in Visual Studio

### PR DESCRIPTION
**Pull request by majocha**
This is related to #14826.

I reshuffled and simplified the old code of `FSharpNavigateToSearchService` quite a bit.

Operators now show up correctly in search results.

The code previously used `Array.Parallel` to collect search matches, so this was essentially not cancellable? This is replaced with `Async.Parallel` which should behave better.

Previously this service used `MemoryCache` with 10 sec lifetime. This may be controversial but I removed the MemoryCache in favor of simply caching everything in a `ConcurrentDictionary`. This could take up maybe a few MBs in worst cases, but there should be a lot less GC and CPU activity on subsequent searches. Cache is now cleared on solution close.

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs [hunk 1]
```diff
@@ -235,53 +118,100 @@ type internal FSharpNavigateToSearchService
...
+    let processDocument (tryMatch: string -> PatternMatch option) (kinds: IImmutableSet<string>) (document: Document) =
+        async {
+            let! sourceText = document.GetTextAsync Async.DefaultCancellationToken |> Async.AwaitTask
+
+            let processItem item =
+                asyncMaybe {
+                    do! Option.guard (kinds.Contains(navigateToItemKindToRoslynKind item.Kind))
+
+                    let name =
+                        // This conversion back from logical names to display names should never be needed, because
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:153]: So then why is this here? :)

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs [hunk 2]
```diff
@@ -235,53 +118,100 @@ type internal FSharpNavigateToSearchService
...
+    let createMatcherFor searchPattern =
+        let patternMatcher =
+            lazy
+                patternMatcherFactory.CreatePatternMatcher(
+                    searchPattern,
+                    PatternMatcherCreationOptions(
+                        cultureInfo = CultureInfo.CurrentUICulture,
+                        flags =
+                            (PatternMatcherCreationFlags.AllowFuzzyMatching
+                             ||| PatternMatcherCreationFlags.AllowSimpleSubstringMatching)
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:132]: Is this a recommended set of flags? Are there any other interesting flags there by chance?

**Review by psfinaki — APPROVED** [review]
Thanks for this. If you have any benchmarks at your hand - feel free to add to the PR.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs (hunk 1 above)
Comment (majocha) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:153]: You wouldn't believe how old this comment is 😢

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs (hunk 2 above)
Comment (majocha) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:132]: fuzzy matching is a must, `AllowSimpleSubstringMatching` probably we could do without, unless it enhances fuzzy matching to substrings? This is very tersely documented unfortunately. 
`IncludeMatchedSpans` would give us that nice highlighting of matching substrings, but we would have to expose this functionality in `IFSharpNavigateToSearchService` on Roslyn side first.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs (hunk 1 above)
Comment (psfinaki) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:153]: Well maybe you can remove all this crap for good now since you are cleaning up anyway :)

Comment (psfinaki) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:153]: But if you have a gut feeling that this should stay - that is also fine.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs (hunk 2 above)
Comment (psfinaki) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:132]: Thanks for looking into this. 

> IncludeMatchedSpans would give us that nice highlighting of matching substrings, but we would have to expose this functionality in IFSharpNavigateToSearchService on Roslyn side first.

Sounds helpful. I think we can do this as a followup and, if so, even bring more attention to this - you are turning this semi-broken functionality into something we can be proud of.

**Review by KevinRansom — APPROVED** [review]

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs (hunk 2 above)
Comment (majocha) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:132]: Actually there is one low hanging fruit. C# lang service has a feature to search by namespace/container like this:
image

I think we could unlock it quite easily for F#.

Comment (psfinaki) [reply vsintegration/src/FSharp.Editor/Navigation/NavigateToSearchService.fs:132]: Yep, sounds good.
